cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+    Percent  int
+    MinCount *int
+    MaxCount *int
+}
+
+type NodeClassBufferCapacity struct {
+    // Upper limit of the total capacity that will be scaled for this buffer.
+    // The total capacity will be split into smaller chunks that each fits onto nodes.
+    // The size of a chunk will be either decided for the user (using some defaults and heuristics) or the
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
+
+    // Optional if defining a single chunk size is needed. The controller will translate this into enough
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector map[string]string
+}
+
+// Subject to change depending on the sig-scheduling work to define workloads and reservations API
+type PodBufferCapacity struct {
+    ReplicaNum int
+    PodSpec    PodSpec
+}
+
+type BufferStatus struct {
+    PodBasedBufferSpec   PodBufferCapacity
+    Conditions           []BufferCondition
+    LastFullyProvisioned *metav1.Time
+}
+```
+
+### What is the responsibility split between the Buffers controller and autoscaler?
+
+The controller would own the logic of translating the buffer configuration into
+a pod spec that would define what spare space needs to be provisioned in the
+cluster to fulfill the requirements of the buffer. The autoscaler component
+would use the pod spec to provision the capacity.
+
+In case of the cluster autoscaler we would deploy the controller as an optional
+subprocess:
+
+![Figure](./images/buffers-component-structure.jpg)
+
+This design has two main advantages compared to embedding all the logic in the
+library that would be used in autoscaler code:
+
+*   Responsibility split: it splits the responsibility of understanding buffer
+    configuration and exposes to the user what exactly will be provisioned in
+    the cluster based on the buffer spec that was provided. Also if for example
+    the target deployment changes it will be visible whether the change was
+    already picked up.
+*   Extendability: this design will allow for the users to either opt in the
+    buffers provided in OSS libraries only or to customize the buffers offering
+    by providing their own logic of translation (with option to use the existing
+    libraries as a base for their own implementation). This way the user could
+    define their custom way of configuring buffers and still use the standard
+    autoscaling tools.
+
+### How to directly define cluster capacity?
+
+Currently the only option to define the cluster capacity is to define a PodSpec.
```

Comment:
Shouldn't this section be updated now that the proposal is sent for review? Also I'm not sure I understand the first sentence, shouldn't it be something like "the only option to consume Node capacity in the cluster is via a Pod"?

Reply:
Thanks! I still think that the use of "defining capacity" here is misleading. Pods/reservations don't define capacity, they consume capacity defined by Nodes/ResourceSlices.